File: articles/john-von-neumann-and-the-architecture-of-modern-computing/java/FetchExecute.java
public class FetchExecute {
    static class Instruction {
        String op;
        int arg;
        Instruction(String op, int arg) { this.op = op; this.arg = arg; }
    }

    public static void main(String[] args) {
        Instruction[] program = {
            new Instruction("LOAD", 2),
            new Instruction("ADD", 3),
            new Instruction("STORE", 0),
            new Instruction("HALT", 0)
        };
        int acc = 0;
        for (Instruction inst : program) {
            switch (inst.op) {
                case "LOAD": acc = inst.arg; break;
                case "ADD": acc += inst.arg; break;
                case "STORE": System.out.printf("store address=%d value=%d%n", inst.arg, acc); break;
                case "HALT": System.out.printf("halt accumulator=%d%n", acc); return;
            }
        }
    }
}

File: articles/john-von-neumann-and-the-architecture-of-modern-computing/cpp/fetch_execute.cpp
#include <iostream>
#include <string>
#include <vector>

struct Instruction {
    std::string op;
    int arg;
};

int main() {
    std::vector<Instruction> program{{"LOAD", 2}, {"ADD", 3}, {"STORE", 0}, {"HALT", 0}};
    int acc = 0;
    for (const auto& inst : program) {
        if (inst.op == "LOAD") acc = inst.arg;
        else if (inst.op == "ADD") acc += inst.arg;
        else if (inst.op == "STORE") std::cout << "store address=" << inst.arg << " value=" << acc << "\n";
        else if (inst.op == "HALT") {
            std::cout << "halt accumulator=" << acc << "\n";
            break;
        }
    }
    return 0;
}
